# Estimating Runs of Homozygosity (RoHs)
First, we split our filtered VCF file into single-sample VCFs, keeping biallelic SNPs. Then, we calculate ROHs with bcftools roh function, setting the allele frequency as 0.4 per default, being thus conservative when identifying ROHs.

```
#!/bin/bash
samples="list/of/samples.txt"
VCF="path/to/VCF/file.vcf.gz"

for i in $(cat "$samples");
do
(
vcftools --indv "$i" --gzvcf "$VCF" \
  --maf 0.01 --min-alleles 2 --max-alleles 2 \
  --remove-indels --minQ 30 --min-meanDP 4 --max-meanDP 150 --minDP 4 --maxDP 150 \
  --max-missing 0.9 --recode --stdout \
  |gzip -c > "$i"_filtered_dataset.vcf.gz
bgzip -d "$i"_filtered_dataset.vcf.gz

bcftools roh -G30 --AF-dflt 0.4 "$i"_filtered_dataset.vcf.gz > results/results_"$i".txt
) &
done
wait
```
