# Time-calibrated SNAPP
To run SNAPP on BEAST2 we used the command `beast -threads <n_threads> snapp_calibrated_viperatemplate.xml` for three different runs. The [template xml](https://github.com/adtalave/EuropeanVipersGenomics/blob/main/snapp_calibrated_viperatemplate.xml) is available in this repository. Afterwards, we checked for the stationarity and convergence of the three runs in Tracer and used TreeAnnotator to construct the maximum clade credibility tree.

# Mitochondrial genomes

To extract mitogenomes from Illumina libraries, we used getorganelle and available mitogenome references from closely-related taxa.
```bash
library="path/to/library"
reference="path/to/reference"
output="output_name"

get_organelle_from_reads.py -1 "$library".fastq.gz -2 "$library".fastq.gz -t 36 -R 10 -k 21,45,65,85,105 -F animal_mt -s "$reference" -o "$output"  
```
Then, the assembled mitogenomes were annotated with mitofinder to extract CDS:
```bash
assembly="path/to/assembled/mitogenome.fasta"
reference="path/to/annotated/reference.gb"

mitofinder -j mitofinder_Vmonsan -a "$assembly" -r "$reference" -o 2 -p <threads> -m <memory>  
```
Fasta sequences were aligned with MEGA11 and muscle algorithm. 

After, we ran partitionfinder2 with the following edited .cfg file and the command `python PartitionFinder.py partition_finder.cfg`.

```
## ALIGNMENT FILE ##
alignment = concatenado.phy;

## BRANCHLENGTHS: linked | unlinked ##
branchlengths = linked;

## MODELS OF EVOLUTION: all | allx | mrbayes | beast | gamma | gammai | <list> ##
models = all;

# MODEL SELECCTION: AIC | AICc | BIC #
model_selection = AICc;

## DATA BLOCKS: see manual for how to define ##
[data_blocks]
ATP6_1stpos = 1-681\3;
ATP6_2ndpos = 2-681\3;
ATP6_3rdpos = 3-681\3;
ATP8_1stpos  = 682-846\3;
ATP8_2ndpos  = 683-846\3;
ATP8_3rdpos  = 684-846\3;
COX1_1stpos  = 847-2448\3;
COX1_2ndpos  = 848-2448\3;
COX1_3rdpos  = 849-2448\3;
COX2_1stpos  = 2449-3141\3;
COX2_2ndpos  = 2450-3141\3;
COX2_3rdpos  = 2451-3141\3;
COX3_1stpos  = 3142-3924\3;
COX3_2ndpos  = 3143-3924\3;
COX3_3rdpos  = 3144-3924\3;
CTYB_1stpos  = 3925-5040\3;
CTYB_2ndpos  = 3926-5040\3;
CTYB_3rdpos  = 3927-5040\3;
ND1_1stpos  = 5041-6003\3;
ND1_2ndpos  = 5042-6003\3;
ND1_3rdpos  = 5043-6003\3;
ND2_1stpos  = 6004-7035\3;
ND2_2ndpos  = 6005-7035\3;
ND2_3rdpos  = 6006-7035\3;
ND3_1stpos  = 7036-7377\3;
ND3_2ndpos  = 7037-7377\3;
ND3_3rdpos  = 7038-7377\3;
ND4_1stpos  = 7378-8091\3;
ND4_2ndpos  = 7379-8091\3;
ND4_3rdpos  = 7380-8091\3;
ND4L_1stpos  = 8092-8382\3;
ND4L_2ndpos  = 8093-8382\3;
ND4L_3rdpos  = 8094-8382\3;
ND5_1stpos  = 8383-10173\3;
ND5_2ndpos  = 8384-10173\3;
ND5_3rdpos  = 8385-10173\3;
ND6_1stpos  = 10174-10695\3;
ND6_2ndpos  = 10175-10695\3;
ND6_3rdpos  = 10176-10695\3;
## SCHEMES, search: all | user | greedy | rcluster | rclusterf | kmeans ##
[schemes]
search = greedy;
```
Model equivalences from partition finder into beast can be found [here](https://justinbagley.rbind.io/2016/10/11/setting-dna-substitution-models-beast/).
.xml file with mitochondrial CDS and the resulting partitions and time constraints found [here](https://github.com/adtalave/EuropeanVipersGenomics/blob/main/mitogenome_beast_13prots_16partitions.xml).

# Introgression with DSuite





# TreeMix


# TWISST


