# Time-calibrated SNAPP
To run SNAPP on BEAST2 we used the command `beast -threads <n_threads> snapp_calibrated_viperatemplate.xml` for three different runs. The [template xml](https://github.com/adtalave/EuropeanVipersGenomics/blob/main/snapp_calibrated_viperatemplate.xml) is available in this repository. Afterwards, we checked for the stationarity and convergence of the three runs in Tracer and used TreeAnnotator to construct the maximum clade credibility tree.

# Mitochondrial genomes

To extract mitogenomes from Illumina libraries, we used getorganelle and available mitogenome references from closely-related taxa.
```bash
library="path/to/library"
reference="path/to/reference"
output="output_name"

get_organelle_from_reads.py -1 "$library".fastq.gz -2 "$library".fastq.gz -t 36 -R 10 -k 21,45,65,85,105 -F animal_mt -s "$reference" -o "$output"  
```
Then, the assembled mitogenomes were annotated with mitofinder to extract CDS:
```bash
assembly="path/to/assembled/mitogenome.fasta"
reference="path/to/annotated/reference.gb"

mitofinder -j mitofinder_Vmonsan -a "$assembly" -r "$reference" -o 2 -p <threads> -m <memory>  
```
Fasta sequences were aligned with MEGA11 and muscle algorithm. 

After, we ran partitionfinder2 with the following edited .cfg file and the command `python PartitionFinder.py partition_finder.cfg`.

```
## ALIGNMENT FILE ##
alignment = concatenado.phy;

## BRANCHLENGTHS: linked | unlinked ##
branchlengths = linked;

## MODELS OF EVOLUTION: all | allx | mrbayes | beast | gamma | gammai | <list> ##
models = all;

# MODEL SELECCTION: AIC | AICc | BIC #
model_selection = AICc;

## DATA BLOCKS: see manual for how to define ##
[data_blocks]
ATP6_1stpos = 1-681\3;
ATP6_2ndpos = 2-681\3;
ATP6_3rdpos = 3-681\3;
ATP8_1stpos  = 682-846\3;
ATP8_2ndpos  = 683-846\3;
ATP8_3rdpos  = 684-846\3;
COX1_1stpos  = 847-2448\3;
COX1_2ndpos  = 848-2448\3;
COX1_3rdpos  = 849-2448\3;
COX2_1stpos  = 2449-3141\3;
COX2_2ndpos  = 2450-3141\3;
COX2_3rdpos  = 2451-3141\3;
COX3_1stpos  = 3142-3924\3;
COX3_2ndpos  = 3143-3924\3;
COX3_3rdpos  = 3144-3924\3;
CTYB_1stpos  = 3925-5040\3;
CTYB_2ndpos  = 3926-5040\3;
CTYB_3rdpos  = 3927-5040\3;
ND1_1stpos  = 5041-6003\3;
ND1_2ndpos  = 5042-6003\3;
ND1_3rdpos  = 5043-6003\3;
ND2_1stpos  = 6004-7035\3;
ND2_2ndpos  = 6005-7035\3;
ND2_3rdpos  = 6006-7035\3;
ND3_1stpos  = 7036-7377\3;
ND3_2ndpos  = 7037-7377\3;
ND3_3rdpos  = 7038-7377\3;
ND4_1stpos  = 7378-8091\3;
ND4_2ndpos  = 7379-8091\3;
ND4_3rdpos  = 7380-8091\3;
ND4L_1stpos  = 8092-8382\3;
ND4L_2ndpos  = 8093-8382\3;
ND4L_3rdpos  = 8094-8382\3;
ND5_1stpos  = 8383-10173\3;
ND5_2ndpos  = 8384-10173\3;
ND5_3rdpos  = 8385-10173\3;
ND6_1stpos  = 10174-10695\3;
ND6_2ndpos  = 10175-10695\3;
ND6_3rdpos  = 10176-10695\3;
## SCHEMES, search: all | user | greedy | rcluster | rclusterf | kmeans ##
[schemes]
search = greedy;
```
Model equivalences from partition finder into beast can be found [here](https://justinbagley.rbind.io/2016/10/11/setting-dna-substitution-models-beast/).
.xml file with mitochondrial CDS and the resulting partitions and time constraints found [here](https://github.com/adtalave/EuropeanVipersGenomics/blob/main/mitogenome_beast_13prots_16partitions.xml).

# Introgression with DSuite
We used the function Dtrios implemented on Dsuite to calculate ABBA-BABA tests with the `--ABBAclustering` flag. 
Then, when a p-value (alpha=0.01) for the ABBA clustering threshold is not significant, o the its |Zscore| < 3 , we set to 0 the value of the f4 branch for that particular trio.


```bash
Dsuite/Build/Dsuite Dtrios --KS-test-for-homoplasy -o <output_name> -t <tree.nwk> <dataset>.vcf.gz SETS.txt
awk -F'\t' -v OFS='\t' 'NR==1 {print $0; next} ($6 > 0.01 || $8 > 0.01 || ($5 >= -3 && $5 <= 3)) {$7=0} 1' <output_name>_tree.txt > sensitive_clustering_<output_name>_tree.txt
#for the robust threshold you can use:
awk -F'\t' -v OFS='\t' 'NR==1 {print $0; next} ($6 > 0.01 || $8 > 0.01 || ($5 >= -3 && $5 <= 3)) {$7=0} 1' <output_name>_tree.txt > robust_clustering_<output_name>_tree.txt
```
Finally, we compute the f-branch estimate for the significant trios and plot it.
```
Dsuite/Build/Dsuite Fbranch <tree>.nwk sensitive_clustering_<output_name>_tree.txt > filtered0.01_robust_<output>_Fbranch.txt
Dsuite/utils/dtools.py filtered0.01_robust_<output>_Fbranch.txt <tree>.nwk
```

# TreeMix
The input file for TreeMix was generated with the script [vcf2treemix.sh](https://github.com/speciationgenomics/scripts/blob/master/vcf2treemix.sh).
```
bash vcf2treemix.sh <dataset>.vcf.gz <output>.clust
```
On the one hand, we calculated with treemix the optimal number of migration events without rooting the trees. We analyzed the data in blocks of 20, 100 and 500 SNPs to account for non-independence of the data, although we already used LD-pruned SNPs. Migration edges were calculated for values of m ranging from 1-10 and the options “-noss” and “-global” activated.

```bash
FILE="<dataset>.treemix.frq.gz"

for m in {1..10}
do
    for i in {1..3}
    do
        case $i in
            1)
                k_value=20
                ;;
            2)
                k_value=100
                ;;
            3)
                k_value=500
                ;;
            *)
                echo "i value out of range"
                exit 1
                ;;
        esac

        treemix -i "$FILE" -m $m -o finaltest."$i"."$m" -root Vlat,Vmon -k "$k_value" -noss -global
    done
done
```
Then, we ran a last rooted run:
```
FILE="<dataset>.treemix.frq.gz"
for i in {0..10}
do
 treemix -i "$FILE" -m $i -o treemix_15sspp_100_$i -root Vlatlat,Vlatgad,Vlataru,Vmonsan,Vmonmon,Vmonatl -k 100 -noss -global > treemix_15sspp_x_${i}_log &
done
```
The optimal number of migrations can be then calulated in R with the package OptM.
```R
library("OptM")
test.optM=optM("ssp_K20,100,500/", orientagraph=F, tsv="results.optM_K20,100,500_allSNPs.tsv", method="Evanno", ignore=NULL, thresh=0.05)
plot_optM(test.optM, method = "Evanno")
```

# TWISST
We first phased haplotypes of a high coverage dataset of all studied species with Beagle.
```
beagle="path/to/beagle.jar"
java -jar $beagle gt=<dataset>.vcf.gz out=phased/phased_Vipera_.vcf.gz nthreads=<n_threads>
```

Then we inferred Neighbour-Joining local trees in non-overlapping 25-kbp windows with PhyML and the script [phyml_sliding_windows.py](https://github.com/simonhmartin/genomics_general), with the requirement of bearing at least 50 SNPs per window.
```
#!/bin/bash
phyml="path/to/script/phyml_sliding_windows.py"
chromosomes="list/of/chromosomes.txt"

for i in $(cat $chromosomes)
do
(
python2.7 $phyml -T <n_threads> -g Vipera45_"$i".geno.gz --windType sites --minSNPs 50 --optimise n --individuals Ctir,CN13667,CN13858,L0,MNCN50498,SPM002443,reference.asp,reference.seo --outgroup Ctir -w 25000 -O 0 -p allspecies_NJ_25kb/phyml_allspp_25kb_"$i" --tmp tmp

) &
done
wait
```
For the resulting local trees (27,217), complete topology weightings were computed by TWISST.

```
python twisst.py -t 25kb.trees.gz --outputTopos outTopo_25kb -w <output>.weights.csv --outgroup cro -g cro Ctir_A,Ctir_B -g seo reference.seo_A,reference.seo_B -g asp reference.asp_A,reference.asp_B -g lat L0_A,L0_B -g amm CN13858_A,CN13858_B -g ber SPM002443_A,SPM002443_B -g mon MNCN50498_A,MNCN50498_B -g urs CN13667_A,CN13667_B --method complete
```
To plot chromosomes with highlighted topologies, we ran this [R script](https://github.com/adtalave/EuropeanVipersGenomics/blob/main/plottingTWISST.R). For that script, TWISST was run independently for each chromosome.
