# Demographic inference with PSMC
For each BAM file, we used mpileup to convert it into a fastq file. `-d` and `-D` should be a third and the double, respectively of the mean coverage of each sample

```bash
#!/bin/bash
reference="path/to/reference/genome.fasta"
samples="samples.txt"

for sample in $(cat $samples)
do
(
bcftools mpileup -C50 -f $refernce "$sample"_mkdup.bam | bcftools call -c - | vcfutils.pl vcf2fq -d <1/3x_coverage> -D <2x_coverage> | gzip > "$sample"_diploid.fq.gz
) &
done
wait
```

We can keep only autosomes.
```
autosomes="list/of/autosomes.txt"
seqtk subseq "$sample"_diploid.fq.gz "$autosomes" > "$sample"_diploid_autosomes.fq
```
We generated psmcfa files from fastqs of each sample.
```bash
psmc/utils/fq2psmcfa -q20 "$sample"_diploid_autosomes.fq > "$sample"_diploid.psmcfa
```
To perform the bootstrapping

```bash
#!/bin/bash

psmc/utils/splitfa "$sample"_diploid.psmcfa > "$sample"_split.psmcfa

for i in $(seq 1 10);
do
psmc -N25 -t15 -r5 -b -p "4+30*2+4+6+10" -o "$sample"_round-"$i".psmc "$sample"_split.psmcfa
done

cat "$sample"_round* "$sample"_diploid.psmcfa > "$sample"_combined_bootstrap.psmc
```
To plot it, we used the script [psmc_plot.pl](https://github.com/lh3/psmc/blob/master/utils/psmc_plot.pl)

```
./psmc_plot.pl -R -u 2.4e-9 -g 6 "$sample"_diploid.psmcfa "$sample"_combined_bootstrap.psmc
```
