# PCA

We can perform the PCA using the VCF with unlinked SNPs generated in 02.VCFfiltering.md

```bash
plink1.9 --vcf unlinked_dataset.vcf.gz --allow-extra-chr --double-id --make-bed --out autosomes.pca
plink --bfile autosomes.pca --double-id --allow-extra-chr --set-missing-var-ids @:# --make-bed --pca --out complete_autosomes.eigen.pca
```
Using the .eigenvec and -eigenval files generated, you can plot the PCA in R.


#Admixture

With the same bed file from the PCA, you can run Admixture with the script [admixture-wrapper.py](https://github.com/dportik/admixture-wrapper).

```
plink --bfile autosomes --recode12 --out autosomes --allow-extra-chr #this creates ped file from the bed file already generated
python admixture-wrapper.py -i folder/with/your/ped/file --kmin 1 --kmax 12 --reps 10 --cv 10 -t 20
```
Results were later plotted with the script [plotADMIXTURE.r](https://github.com/speciationgenomics/scripts/blob/master/plotADMIXTURE.r).
