# Differential gene expression analysis
Filtered RNA-Seq reads were mapped to V. latastei reference genome with STAR. STAR needs an index before mapping.

```bash
reference="path/to/your/reference.fasta"
annotation="path/to/your/reference.gtf"
index_dir="path/to/genome/index/folder/"
library="library/prefix"


echo "STAR indexing genome"
STAR-2.7.10b/bin/Linux_x86_64/STAR --runThreadN <n_threads> --runMode genomeGenerate \
  --genomeDir "$index_dir" \
  --genomeFastaFiles "$reference" \
  --sjdbGTFfile "$annotation"

echo "Starting RNA-Seq mapping"

STAR-2.7.10b/bin/Linux_x86_64/STAR --genomeDir "$index_dir" \
  --runThreadN <n_threads> \
  --readFilesIn "$library"_1_fastp2.fastq "$library"_2_fastp2.fastq \
  --outFileNamePrefix "$library" \
  --outSAMtype BAM SortedByCoordinate \
  --outSAMunmapped Within \
  --outSAMattributes Standard

samtools index "$library"Aligned.sortedByCoord.out.bam

```


To quality-check the mapping:
```
gtfToGenePred -genePredExt "$annotation" reference_annotation.refFlat.txt
cat reference_annotation.refFlat.txt | awk '{print $12"\t"$0}' | cut -d$'\t' -f1-11 > tmp.txt
mv tmp.txt reference_annotation.refFlat.txt
java -jar picard/picard.jar CollectRnaSeqMetrics I=$libraryAligned.sortedByCoord.out.bam O=qc_alignment/$library.output.RNA_Metrics REF_FLAT=reference_annotation.refFlat.txt STRAND=NONE 
```

Later we counted gene hits:
```
#!/bin/bash
htseq-count -m intersection-strict -f bam -s yes -r pos --additional-attr "gene_name" "$library"Aligned.sortedByCoord.out.bam $annotation.gff > htseq-counts/"$library".htseq-counts
```

Afterwards, we used the DESeq2 package in R, following the script [07.2.DESeq2.R](https://github.com/adtalave/EuropeanVipersGenomics/blob/main/07.2.DESeq2.R) to proceed with the Differential Gene Expression Analysis and produce a Heatmap of gene expression.

